lib/galaxy/webapps/galaxy/api/jobs.py

```diff
@@ -138,19 +167,43 @@ def build_and_apply_filters(query, objects, filter_func):
         query = build_and_apply_filters(query, kwd.get('date_range_max', None), lambda dmax: Job.table.c.update_time <= dmax)
 
         history_id = kwd.get('history_id', None)
-        if history_id is not None:
-            try:
+        workflow_id = kwd.get('workflow_id', None)
+        invocation_id = kwd.get('invocation_id', None)
+        try:
+            if history_id is not None:
                 decoded_history_id = self.decode_id(history_id)
                 query = query.filter(Job.table.c.history_id == decoded_history_id)
-            except Exception:
-                raise exceptions.ObjectAttributeInvalidException()
+            if workflow_id is not None:
+                decoded_workflow_id = self.decode_id(workflow_id)
+                query = query.filter(and_(
+                    Job.table.c.id == model.WorkflowInvocationStep.table.c.job_id,
```

Comment:
Would you be interested in adding a query that would also include the implicit collection jobs here ?

As it is now the filter skips jobs that are part of a workflow invocation and that are "mapped over", meaning a single step that creates many jobs. If you want to include these you would need to select jobs that are part of a ImplicitCollectionJobsJobAssociation model. Let me know if you need help with this.

Reply:
Yes, I would need some help with this.

How is ImplicitCollectionJobsJobAssociation related to a workflow? What are the attributes that we would care about? Would it be `implicit_collection_jobs_id` that have a matching `job_id` to jobs associated with the workflow? I am not sure.

I thought that maybe

```
query = query.filter(and_(
    Job.table.c.id == model.WorkflowInvocationStep.table.c.job_id,
    model.WorkflowInvocationStep.table.c.workflow_invocation_id == model.WorkflowInvocation.table.c.id,
    model.WorkflowInvocation.table.c.workflow_id == decoded_workflow_id
))
```

could be replaced with

```
query = query.filter(and_(
    Job.table.c.id == model.WorkflowInvocationStep.table.c.job_id,
    model.WorkflowInvocationStep.table.c.workflow_invocation_id == model.WorkflowInvocation.table.c.id,
    model.WorkflowInvocation.table.c.workflow_id == decoded_workflow_id
))
query = query.outerjoin(
    model.ImplicitCollectionJobsJobAssociation,
    Job.table.c.id == model.ImplicitCollectionJobsJobAssociation.table.c.job_id
)
```

but the resulting query printed with `print(str(query))` does not seem to select any attributes from ImplicitCollectionJobsJobAssociation.

Reply:
@rikeshi if you invoke a workflow with a collection as input, it creates an implicit job for the collection as well as individual jobs for each collection element. The former are missed by the query at the moment. As far as I understand you would need to get implicit_collection_jobs_id from the WorkflowInvocationStep table and then find the job_id which matches this in the ImplicitCollectionJobsJobAssociation table.

(Please correct if something here was inaccurate @mvdbeek.)

We can talk about this later on today if you like.

Reply:
@simonbray I have changed the code according to your comment. But I am not positive that it is correct. Does it look OK to you?

Reply:
I tested locally and the results are not quite correct. Maybe you can have a look at this and cherry-pick if it looks ok to you? https://github.com/galaxyproject/galaxy/commit/0f64a66429f26fe0a4e51bcaffbe8b22f9e00f1f